config/policy.go

```diff
@@ -0,0 +1,254 @@
+	HTTPEndpointOption
+	TLSEndpointOption
+	TCPEndpointOption
+	JsonConvertible
+} {
+	cfg := &policy{}
+	applyOpts(cfg, opts...)
+
+	return cfg
+}
+
+// WithInboundRules adds the provided policy rules to the inbound
+// set of the given policy.
+// The order in which policies are specified is observed.
+func WithInboundRules(opts ...policyRuleSetOption) policyOption {
+	rules := []policyRule{}
+	applyOpts(&rules, opts...)
+
+	return inboundRules(rules)
+}
+
+// WithOutboundRules adds the provided policy to be outbound
+// set of the given policy.
+// The order in which policies are specified is observed.
+func WithOutboundRules(opts ...policyRuleSetOption) policyOption {
+	rules := []policyRule{}
+	applyOpts(&rules, opts...)
+
+	return outboundRules(rules)
+}
+
+// WithPolicyRule provides a policy rule built from the given options.
+func WithPolicyRule(opts ...policyRuleOption) interface {
+	policyRuleSetOption
+	JsonConvertible
+} {
+	pr := policyRule{}
+	applyOpts(&pr, opts...)
+
+	return pr
+}
+
+// WithPolicyName sets the provided name on a policy rule.
+func WithPolicyName(n string) policyRuleOption {
+	return optionFunc[*policyRule](
+		func(r *policyRule) {
+			r.Name = n
+		})
+}
+
+// WithPolicyExpression appends the provided CEL statement to a policy rule's expressions.
+func WithPolicyExpression(e string) policyRuleOption {
+	return optionFunc[*policyRule](
+		func(r *policyRule) {
+			r.Expressions = append(r.Expressions, e)
+		})
+}
+
+// WithPolicyAction appends the provided action to the set of the policy rule.
+// The order the actions are specified is observed.
+func WithPolicyAction(opts ...policyActionOption) interface {
+	policyRuleOption
+	JsonConvertible
+} {
+	a := action{}
+	applyOpts(&a, opts...)
+
+	return a
+}
+
+// WithActionType sets the provided type for this action. Type must be specified.
+func WithPolicyActionType(t string) policyActionOption {
+	return optionFunc[*action](func(a *action) { a.Type = t })
+}
+
+// WithConfig sets the provided map as the configuration for this action
+func WithPolicyActionConfig(c map[string]any) policyActionOption {
+	return optionFunc[*action](
+		func(a *action) {
+			raw, err := json.Marshal(c)
+			if err != nil {
+				panic("unable to marshal action config")
+			}
+			a.Config = raw
+		})
+}
+
+// supports inbound rules as an a policy option
+func (ib inboundRules) apply(p *policy) {
+	p.Inbound = append(p.Inbound, ib...)
+}
+
+// supports outbound rules as a policy option
+func (ib outboundRules) apply(p *policy) {
+	p.Outbound = append(p.Outbound, ib...)
+}
+
+// supports policy rule as an option of a collection of
+// rules, which can be used for inbound or outbound
+func (pr policyRule) apply(r *[]policyRule) {
+	*r = append(*r, pr)
+}
+
+// supports action as an option of a policy rule
+func (a action) apply(p *policyRule) {
+	p.Actions = append(p.Actions, a)
+}
+
+// an option that is applicable to the specified type
+type option[T any] interface {
+	apply(T)
+}
```

Comment:
Using this interface allowed me to have policyRule and action be both structs that can be converted to json _and_ be 'options' of policy and policyRule respectively. I can't tell if I got too "clever" with it to the point that it's not easily follow-able.